File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.cpp
/*
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

Node* newNode(int val) {
    return new Node(val);
}
*/

class Solution {
  public:
    int diameter(Node* root) {
        // Your code here
        int diameter=0;
        height(root,diameter);
        return diameter;
    }
    
    private:
      int height(Node* root, int& diameter){
        if(root==NULL) return 0;
        int lh =height(root->left,diameter);
        int rh =height(root->right,diameter);
        diameter = max(diameter,lh+rh);
        return 1+max(lh,rh);
    }
    
    
};

File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.py
'''
# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
class Solution:
    def diameter(self, root):
        # Your code here
        self.dia=0
        def height(root):
            if root is None:
                return 0;
                
            lh =  height(root.left)
            rh = height(root.right)
            self.dia = max(self.dia,lh+rh)
            return 1 + max(lh, rh)
        height(root)    
        return self.dia
        
        